source/common/network/connection_impl.cc

```diff
@@ -174,19 +176,24 @@ void ConnectionImpl::readDisable(bool disable) {
   UNREFERENCED_PARAMETER(read_enabled);
   conn_log_trace("readDisable: enabled={} disable={}", *this, read_enabled, disable);
 
-  // We do not actually disable reading from the socket. We just stop firing read callbacks.
-  // This allows us to still detect remote close in a timely manner. In practice there is a chance
-  // that a bad client could send us a large amount of data on a HTTP/1.1 connection while we are
-  // processing the current request.
+  // When we disable reads, we still allow for early close notifications (the equivalent of
+  // EPOLLRDHUP for an epoll backend). For backends that support it, this allows us to apply
+  // back pressure at the kernel layer, but still get timely notification of a FIN. Note that
+  // we are not gaurenteed to get notified, so even if the remote has closed, we may not know
+  // until we try to write. Further note that currently we don't correctly handle half closed
+  // TCP connections in the sense that we assume that a remote FIN means the remote intends a
+  // full close.
   //
-  // TODO(mattklein123): Add buffered data stats and potentially fail safe processing that
-  // disconnects or applies back pressure to bad HTTP/1.1 clients.
+  // TODO(mattklein123): Potentially support half-closed TCP connections. It's unclear if this is
+  // required for any scenarios in which Envoy will be used (I don't know of any).
   if (disable) {
     ASSERT(read_enabled);
     state_ &= ~InternalState::ReadEnabled;
+    file_event_->setEnabled(Event::FileReadyType::Write | Event::FileReadyType::Closed);
   } else {
     ASSERT(!read_enabled);
     state_ |= InternalState::ReadEnabled;
+    file_event_->setEnabled(Event::FileReadyType::Read | Event::FileReadyType::Write);
```

Comment:
Why don't we listen for close when read enabled? Is this because EPOLLRDHUP is reflected as a read? If so, a comment would be helpful.